### ISSUE TITLE
Operator issue with sqlalchemy.func

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#7743]
### Discussed in https://github.com/microsoft/pyright/discussions/7742

<div type='discussions-op-text'>

<sup>Originally posted by **Apakottur** April 22, 2024</sup>
Consider the following file:
```python
# test.py
from sqlalchemy import func
import datetime as dt

reveal_type(func.current_date() + dt.timedelta(days=1))
```

Running `mypy test.py` gives:
```text
test.py:5: note: Revealed type is "sqlalchemy.sql.elements.ColumnElement[datetime.date]"
Success: no issues found in 1 source file
```

Running `pyright test.py` gives:
```text
test.py
  test.py:5:13 - error: Operator "+" not supported for types "current_date" and "timedelta" (reportOperatorIssue)
  test.py:5:13 - information: Type of "func.current_date() + dt.timedelta(days=1)" is "Unknown"
1 error, 0 warnings, 1 information
```

The code works as expected, and I would expect pyright to not raise on error on it.
Is it a bug in pyright or an incorrect typing with sqlalchemy?

Versions:
python 3.12
sqlalchemy 2.0.29
pyright 1.1.359
mypy 1.9.0</div>

**Comment by erictraut** [comment]
Here's a simplified code sample:

```python
from datetime import timedelta
from sqlalchemy.sql.functions import current_date

def func(a: current_date, b: timedelta):
    reveal_type(a.__add__) # Appears not to bind correctly
    a.__add__(b)

    return a + b
```

**Comment by Azureblade3808** [comment]
It seems that there is a missing overload in SQLAlchemy.

For `sqlalchemy.sql.elements.SQLCoreOperations`, its `__add__` method is declared as such --
```python
        @overload
        def __add__(
            self: _SQO[_NMT],
            other: Any,
        ) -> ColumnElement[_NMT]: ...

        @overload
        def __add__(
            self: _SQO[str],
            other: Any,
        ) -> ColumnElement[str]: ...

        def __add__(self, other: Any) -> ColumnElement[Any]: ...
```
, whereas `_SQO` is an alias of `SQLCoreOperations` and `_NMT` is a type variable with a bound of `float`, `int` or `Decimal`. We can see that type of `self` in overloads doesn't cover everything including `SQLCoreOperations[datetime.datetime]` in the sample.

If the code above is modified with an additional overload --
```python
        @overload
        def __add__(
            self: _SQO[_NMT],
            other: Any,
        ) -> ColumnElement[_NMT]: ...

        @overload
        def __add__(
            self: _SQO[str],
            other: Any,
        ) -> ColumnElement[str]: ...

        # Added.
        @overload
        def __add__(self, other: Any) -> ColumnElement[Any]: ...

        def __add__(self, other: Any) -> ColumnElement[Any]: ...
```
, no error gets reported.

But I think the error message could be improved and the method should be uncallable in its current form.

**Comment by erictraut** [comment]
Good analysis. I agree that the error message could be improved. I also think `a.__add__` should generate an error since `__add__` cannot be bound to `a`.

**Comment by erictraut** [comment]
@Apakottur, you asked whether this was a bug in the sqlalchemy stubs or in pyright. It appears to be a bit of both. I recommend filing a separate bug for the stubs issue.

**Comment by Apakottur** [comment]
Opened a PR to fix the types, if it wont get accepted I'll open an issue.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.360, which I just published.
